### PR TITLE
Add `VerifyTx` to `executor.Manager`

### PR DESCRIPTION
**Pull request by dhrubabasu**
## Why this should be merged

Clearer separation of concerns between the `Manager` and `Builder` in platformvm. Mirrors the AVM implementation: https://github.com/ava-labs/avalanchego/blob/v1.10.15/vms/avm/block/executor/manager.go#L42-L44

Factored out of https://github.com/ava-labs/avalanchego/pull/2296

## How this works

pretty straightforward

## How this was tested

CI

### CODE REVIEW

**Review by danlaine — COMMENTED** [review]
I think I prefer the existing code. All else being equal, I like to keep interfaces smaller, and to me, `VerifyTx` doesn't feel natural to put on the block execution manager. All the other methods deal with blocks, not transactions.

**Comment by dhrubabasu** [comment]
> I think I prefer the existing code. All else being equal, I like to keep interfaces smaller, and to me, `VerifyTx` doesn't feel natural to put on the block execution manager. All the other methods deal with blocks, not transactions.

The manager manages the `preferred` state so it seemed fitting to me that it should also manage verification of transactions against that preferred state. The end state I'm trying to get to w.r.t. separation of concerns is here (from #2296): https://github.com/ava-labs/avalanchego/blob/remove-add-unverified-tx/vms/platformvm/network/network.go#L162-L181

The AVM follows this pattern too and mostly trying to get alignment between the two vms where it makes sense.

File: vms/platformvm/block/executor/manager.go [hunk 1]
```diff
@@ -103,3 +116,16 @@ func (m *manager) SetPreference(blockID ids.ID) (updated bool) {
 func (m *manager) Preferred() ids.ID {
 	return m.preferred
 }
+
+func (m *manager) VerifyTx(tx *txs.Tx) error {
```
Comment (abi87) [inline vms/platformvm/block/executor/manager.go:120]: what I think it's a bit confusing here is that VerifyTx is to verify mempool transactions but we already have a MempoolTxVerifier visitor that basically does all of it already.
So maybe the change to be done is to move the "isBootstrapped" check into the mempoolVerified and that's it.
This would avoid increasing the blkManager interface (as per @danlaine 's comment) and make blkManager more of a backend object for verification of both mempool txs and txs in block.

**Review by abi87 — CHANGES_REQUESTED** [review]
jsut move "isBootstrap" change into MempoolTxVerified without introducing VerifyTx

File: vms/platformvm/block/builder/builder.go [hunk 2]
```diff
@@ -103,24 +102,14 @@ func New(
 
 // AddUnverifiedTx verifies a transaction and attempts to add it to the mempool
 func (b *builder) AddUnverifiedTx(tx *txs.Tx) error {
-	if !b.txExecutorBackend.Bootstrapped.Get() {
-		return ErrChainNotSynced
-	}
-
 	txID := tx.ID()
 	if b.Mempool.Has(txID) {
```
Comment (abi87) [inline vms/platformvm/block/builder/builder.go:106]: honestly also this mempool.Has check could be in MempoolTxVerifier. We could evolve the MempoolTxVerifier to include the full mempool, so to be able to carry out all relevant checks (including conflicts on zero confimation utxo expenditure in P-chain mempool, if we ever want to enable it)

File: vms/platformvm/block/executor/manager.go (hunk 1 above)
Comment (dhrubabasu) [reply vms/platformvm/block/executor/manager.go:120]: This function uses the `MempoolTxVerifier`. I don't think it's a good design pattern to required consumers of this package create a struct themselves to then call it. I think this pattern (https://github.com/ava-labs/avalanchego/blob/remove-add-unverified-tx/vms/platformvm/network/network.go#L162-L181) is much more digestible:

```golang
	// Verify the tx at the currently preferred state
	if err := n.manager.VerifyTx(tx); err != nil {
		n.ctx.Log.Debug("tx failed verification",
			zap.Stringer("txID", txID),
			zap.Error(err),
		)

		n.mempool.MarkDropped(txID, err)
		return err
	}
```

than the existing:
```golang
	verifier := txexecutor.MempoolTxVerifier{
		Backend:       b.txExecutorBackend,
		ParentID:      b.blkManager.Preferred(), // We want to build off of the preferred block
		StateVersions: b.blkManager,
		Tx:            tx,
	}
	if err := tx.Unsigned.Visit(&verifier); err != nil {
		b.MarkDropped(txID, err)
		return err
	}
```

File: vms/platformvm/block/builder/builder.go (hunk 2 above)
Comment (dhrubabasu) [reply vms/platformvm/block/builder/builder.go:106]: I'd like to defer making MempoolTxVerifier more bloated than it needs to be- this check is being moved to `network.issueTx` in #2296. This PR is a stepping stone towards that

**Comment by abi87** [comment]
> I think I prefer the existing code. All else being equal, I like to keep interfaces smaller, and to me, `VerifyTx` doesn't feel natural to put on the block execution manager. All the other methods deal with blocks, not transactions.

TL;DR I am fine approving this PR if we change `blkManager.VerifyTx` to something like `blkManager.VerifyMempoolCandidateTx`, *both in the P-chain and the X-chain*.


`blkManager.VerifyTx` is a misleading name. I dived into X-chain mempool and  found out that what `VerifyTx` is doing is not really verifying a generic tx, but only checking that mempool-candidate txs verify against chain state.
`VerifyTx` is *not* used to verify transactions in a block, not even in the X-chain, which is why I think the name sucks.

@danlaine we could discuss whether blkManager is the right "actor" to carry out state verification of mempool-candidate txs.
I think it's not optimal. But I can accept this PR as an improvement while we look for a cleaner design.

The reason is that X-chain mempool does a better job at validating txs from rejected blocks.
P-chain does not really do that, which is why we only needed state-verifications against mempool candidate in the blockBuilder.
This PR is a step towards hardening P-chain mempool.
I'd suggest that we can discuss a better and possibly shared design for all of this once we harden P-chain mempool.

File: vms/platformvm/block/builder/builder.go [hunk 3]
```diff
@@ -103,24 +102,14 @@ func New(
...
 	}
 
-	verifier := txexecutor.MempoolTxVerifier{
-		Backend:       b.txExecutorBackend,
-		ParentID:      b.blkManager.Preferred(), // We want to build off of the preferred block
-		StateVersions: b.blkManager,
-		Tx:            tx,
-	}
-	if err := tx.Unsigned.Visit(&verifier); err != nil {
+	if err := b.blkManager.VerifyTx(tx); err != nil {
```
Comment (abi87) [inline vms/platformvm/block/builder/builder.go:112]: Rename `VerifyTx` to something like `VerifyMempoolCandidate` both here and in X-chain.
This way we can avoid confusion and minimize awkwardness of having to state verify mempool candidate in blkManager. We can fully solve the awkwardness later on and jontly across P-chain and X-chain

File: vms/platformvm/block/executor/manager.go (hunk 1 above)
Comment (abi87) [reply vms/platformvm/block/executor/manager.go:120]: I don't really think it's more digestible but I'd suggest we find a proper name for `manager.VerifyTx` (my suggestion being `manager.VerifyMempoolCandidate`) and we defer discussions around mempool refactoring.

**Review by abi87 — CHANGES_REQUESTED** [review]
rename blkManager.VerifyTx

**Review by danlaine — APPROVED** [review]

**Review by abi87 — APPROVED** [review]

**Review by StephenButtolph — APPROVED** [review]
